### DISCUSSION TITLE
How to call `run_system_once_with`?

### DISCUSSION THREAD

**Discussion by bbarker** [bevyengine/bevy#11731] [Q&A] [answered]
Say I have a system with function signature:

```rust
pub fn get_player_no_ecs(
    query: Query<(Entity, &Name), With<Player>>,
    player_name: impl Into<String>,
) -> Option<Entity> 
```

Looking at the [system piping example](https://github.com/bevyengine/bevy/blob/main/examples/ecs/system_piping.rs), I thought maybe I should do something like this:

```rust
pub fn get_player_no_ecs(
    query: Query<(Entity, &Name), With<Player>>,
    In(player_name): In<impl Into<String>>,
) -> Option<Entity> 
```

However, I'm not sure how to use this with `World::run_system_once_with` (supposing I have `player_name` in hand at the call site).

**Comment by NiklasEi** [comment]
It should "just" be
```rust
world.run_system_once_with(player_name, get_player_no_ecs)
```
are you running into issues with that?

I do something [very similar in Cinnog](https://github.com/NiklasEi/cinnog/blob/3490b520013f460b5fdf6632767af594fa2f4949/src/lib.rs#L43). It looks like `run_system_once_with` could save me some code there actually :thinking:

**Reply by bbarker** [reply]
I can get past the issue if I remove the string parameter (`player_name`) entirely - of course I actually need it in the game logic.               

Otherwise, I've not been able to figure out why my system fn doesn't satisfy the trait bounds yet:

```
error[E0277]: the trait bound `for<'a, 'b, 'c> fn(bevy::prelude::Query<'a, 'b, (bevy::prelude::Entity, &'c bevy::prelude::Name), bevy::prelude::With<components::Player>>, bevy::prelude::In<_>) -> std::option::Option<bevy::prelude::Entity> {get_player_no_ecs::<_>}: bevy::prelude::IntoSystem<bevy::prelude::In<impl Into<String>>, _, _>` is not satisfied
   --> src/player.rs:370:47
    |
370 |     ecs.run_system_once_with(In(player_name), get_player_no_ecs)
    |         --------------------                  ^^^^^^^^^^^^^^^^^ the trait `bevy::prelude::IntoSystem<bevy::prelude::In<impl Into<String>>, _, _>` is not implemented for fn item `for<'a, 'b, 'c> fn(bevy::prelude::Query<'a, 'b, (bevy::prelude::Entity, &'c bevy::prelude::Name), bevy::prelude::With<components::Player>>, bevy::prelude::In<_>) -> std::option::Option<bevy::prelude::Entity> {get_player_no_ecs::<_>}`
    |         |
    |         required by a bound introduced by this call
    |
note: required by a bound in `run_system_once_with`
   --> /home/bbarker/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_ecs-0.12.1/src/system/system.rs:265:32
    |
265 | ...m_once_with<T: IntoSystem<In, Out, Marker>, In, Out, Marker>(
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RunSystemOnce::run_system_once_with`

```

**Reply by NiklasEi** [reply, marked as answer]
Is the `In` parameter the [first Parameter of the system](https://github.com/bevyengine/bevy/blob/9f2eabb02f726144dbe89644f667ee0a4da88708/crates/bevy_ecs/src/system/mod.rs#L218)?

**Reply by bbarker** [reply]
ah, that was it! I thought I tried that at some point, but probably not in the right permutation with usage of `In`
